nntrainer/tensor/blas_neon.cpp

```diff
@@ -278,250 +266,365 @@ void sgemv_transpose_neon(const float *A, const float *X, float *Y,
+
+    vst1q_f32(&Y32[idx], y0_3);
+    vst1q_f32(&Y32[idx + 4], y4_7);
+  }
+  for (; rows - idx >= 4; idx += 4) {
+    float32x4_t y0_3_32 = vcvt_f32_f16(vld1_f16(&Y[idx]));
+    y0_3_32 = vmulq_f32(y0_3_32, v_beta_32);
 
-  for (unsigned int i = 0; i < rows; i += 8) {
-    float16x8_t y = vld1q_f16(&Y[i]);
-    y = vmulq_f16(v_beta, y);
-    vst1q_f16(&Y[i], y);
+    vst1q_f32(&Y32[idx], y0_3_32);
+  }
+  while (idx < rows) {
+    Y32[idx] = Y[idx] * beta;
+    ++idx;
   }
 
-  float16x8_t v_alpha = vmovq_n_f16(alpha);
+  idx = 0;
+  for (; cols - idx >= 32; idx += 32) {
+    float32x4_t x0_3_f32 = vcvt_f32_f16(vld1_f16(&X[idx]));
+    float32x4_t x4_7_f32 = vcvt_f32_f16(vld1_f16(&X[idx + 4]));
+    float32x4_t x8_11_f32 = vcvt_f32_f16(vld1_f16(&X[idx + 8]));
+    float32x4_t x12_15_f32 = vcvt_f32_f16(vld1_f16(&X[idx + 12]));
+    float32x4_t x16_19_f32 = vcvt_f32_f16(vld1_f16(&X[idx + 16]));
+    float32x4_t x20_23_f32 = vcvt_f32_f16(vld1_f16(&X[idx + 20]));
+    float32x4_t x24_27_f32 = vcvt_f32_f16(vld1_f16(&X[idx + 24]));
+    float32x4_t x28_31_f32 = vcvt_f32_f16(vld1_f16(&X[idx + 28]));
+
+    if (alpha != 1.0) {
+      x0_3_f32 = vmulq_n_f32(x0_3_f32, alpha);
+      x4_7_f32 = vmulq_n_f32(x4_7_f32, alpha);
+      x8_11_f32 = vmulq_n_f32(x8_11_f32, alpha);
+      x12_15_f32 = vmulq_n_f32(x12_15_f32, alpha);
+      x16_19_f32 = vmulq_n_f32(x16_19_f32, alpha);
+      x20_23_f32 = vmulq_n_f32(x20_23_f32, alpha);
+      x24_27_f32 = vmulq_n_f32(x24_27_f32, alpha);
+      x28_31_f32 = vmulq_n_f32(x28_31_f32, alpha);
+    }
 
-  if (cols % 32 == 0) {
-    for (unsigned i = 0; i < cols; i += 32) {
-      float16x8_t x0_7 = vld1q_f16(&X[i]);
-      float16x8_t x8_15 = vld1q_f16(&X[i + 8]);
-      float16x8_t x16_23 = vld1q_f16(&X[i + 16]);
-      float16x8_t x24_31 = vld1q_f16(&X[i + 24]);
+    const __fp16 *__restrict w;
+
+    float yVal;
+
+    for (unsigned int j = 0; j < rows; ++j) {
+      w = &A[j * cols + idx];
+      float32x4_t wvec0_3_f32 = vcvt_f32_f16(vld1_f16(&w[0]));
+      float32x4_t wvec4_7_f32 = vcvt_f32_f16(vld1_f16(&w[4]));
+      float32x4_t wvec8_11_f32 = vcvt_f32_f16(vld1_f16(&w[8]));
+      float32x4_t wvec12_15_f32 = vcvt_f32_f16(vld1_f16(&w[12]));
+      float32x4_t wvec16_19_f32 = vcvt_f32_f16(vld1_f16(&w[16]));
+      float32x4_t wvec20_23_f32 = vcvt_f32_f16(vld1_f16(&w[20]));
+      float32x4_t wvec24_27_f32 = vcvt_f32_f16(vld1_f16(&w[24]));
+      float32x4_t wvec28_31_f32 = vcvt_f32_f16(vld1_f16(&w[28]));
+
+      float32x4_t y0 = vmulq_f32(wvec0_3_f32, x0_3_f32);
+      y0 = vfmaq_f32(y0, wvec4_7_f32, x4_7_f32);
+      y0 = vfmaq_f32(y0, wvec8_11_f32, x8_11_f32);
+      y0 = vfmaq_f32(y0, wvec12_15_f32, x12_15_f32);
+      y0 = vfmaq_f32(y0, wvec16_19_f32, x16_19_f32);
+      y0 = vfmaq_f32(y0, wvec20_23_f32, x20_23_f32);
+      y0 = vfmaq_f32(y0, wvec24_27_f32, x24_27_f32);
+      y0 = vfmaq_f32(y0, wvec28_31_f32, x28_31_f32);
+
+      yVal = vaddvq_f32(y0);
+
+      Y32[j] += yVal;
```

Comment:
```suggestion
      Y32[j] += vaddvq_f32(y0);
```
would this work? (removing all `float yVal`)

Reply:
Good point. I will check and revise if it works.